public/tutorials/debugging-a-smart-contract.md

```diff
@@ -0,0 +1,621 @@
+
+   ```shell
+   { [String: '0'] s: 1, e: 0, c: [ 0 ] }
+   ```
+
+   The important information here is the value `0`. Variables with integer types in Solidity are automatically populated with the value of zero, unlike other languages where it might be `NULL` or `undefined`.
+
+   So this shows us that our variable, `myVariable` is set to `0`.
+
+   <p class="alert alert-info">
+   <strong>Note</strong>: Even though the response shows the `0` listed as a type `String`, the value in the variable is an integer as we specified.
+   </p>
+
+1. Now let's run a transaction on our contract. We'll do this by running the `set()` function, where we can set our variable value to some other integer. Run the following command:
+
+   ```shell
+   SimpleStorage.deployed().then(function(instance){return instance.set(4);});
+   ```
+
+   This sets the variable to `4`. The output shows some information about the transaction:
+
+   ```shell
+    { tx: '0x7f799ad56584199db36bd617b77cc1d825ff18714e80da9d2d5a0a9fff5b4d42',
+      receipt:
+       { transactionHash: '0x7f799ad56584199db36bd617b77cc1d825ff18714e80da9d2d5a0a9fff5b4d42',
+         transactionIndex: 0,
+         blockHash: '0x60adbf0523622dc1be52c627f37644ce0a343c8e7c8955b34c5a592da7d7c651',
+         blockNumber: 5,
+         gasUsed: 41577,
+         cumulativeGasUsed: 41577,
+         contractAddress: null,
+         logs: [] },
+      logs: [] }   
+   ```
+
+   Most important to us is the transaction ID (listed both as `tx` and as `transactionHash`). We'll need to copy that value when we start to debug.
+
+1. To verify that the variable has changed values, run the `get()` function again:
+
+   ```shell
+   SimpleStorage.deployed().then(function(instance){return instance.get.call();});
+   ```
+
+   The output should look like this:
+
+   ```shell
+   { [String: '4'] s: 1, e: 0, c: [ 4 ] }
+   ```
+
+
+## Debugging errors
+
+The above shows how the contract *should* work. **Now, we will introduce some small errors to the contract** and redeploy it. We will see how the issues present itself, and also **see how we can use Truffle's built-in debug feature to fix it**.
+
+We will look at the following issues:
+
+* An infinite loop
+* Invalid error check 
+* No error, but a function isn't operating as desired
+
+
+### Issue #1: An infinite loop
+
+On the Ethereum blockchain, transactions cannot be set to run forever.
+
+A transaction can run up until its gas limit is reached. Once that happens, the transaction will error out, and an "out of gas" error will be returned.
+
+Since gas is priced in Ether, this could have real-world financial implications. So fixing an out-of-gas error is critical.
+
+#### Introducing the error
+
+An infinite loop is easy to create.
+
+1. Open the `Store.sol` file in your `/contracts` directory in a text editor.
+
+1. Replace the `set()` function with the following:
+
+   ```shell
+   function set(uint x) {
+     while(true){myVariable = x;}
```

Comment:
I'd split the while into 3 lines for clarity:

```javascript
while (true) {
  myVariable = x;
}
```

Reply:
Nice. I think that makes it easier to read.